### PR TITLE
Fix storing test results of non-tested schemes

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/6437

### Short description 📝

Until now, we didn't have a way to save state in the mappers. We introduced two ways to get around this limitation:
- Mutating non-mapper objects
- Using file system as a way to store state

Both of these solutions break the functional nature of the mappers.

Instead, we are introducing a new `MapperEnvironment` that's both an input and output of the graph mappers (but can be introduced to non-graph mappers as well):
```swift
func map(graph: Graph, environment: MapperEnvironment) async throws -> (Graph, [SideEffectDescriptor], MapperEnvironment)
```

The `MapperEnvironment` stores data in a dictionary accessed by custom keys that can be defined where needed, similar to SwiftUI's Environment (and the [TCA environment](https://github.com/tgrapperon/swift-composable-environment?tab=readme-ov-file#defining-dependencies)).

Example:
```swift
private struct InitialGraphKey: MapperEnvironmentKey {
    static var defaultValue: Graph?
}

extension MapperEnvironment {
    public var initialGraph: Graph? {
        get { self[InitialGraphKey.self] }
        set { self[InitialGraphKey.self] = newValue }
    }
}
```

In this PR, we're removing `AutomationStorage` and saving test hashes to the disk. Later on, we should also remove the `CacheAnalyticsStore` in favor of returning the analytics directly from the mappers via the environment.

On top of this, you can notice a bunch of changes in `TestService` – this fixes an issue with selective testing in the following scenario:
```sh
tuist test TestsA # Hashes for the _whole_ graph are saved here!
tuist test TestsB # TestsB is skipped even though we have not run tests for it
```

That is if we run tests for a specific scheme, we still save hashes for the whole graph as if we ran the tests for all targets. Instead, we now save tests only for those targets that we actually test.

### How to test the changes locally 🧐

CI should pass. The above scenario must be tested in the private repo that contains the core selective testing logic.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x]  In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistCore/Graph/Mappers/MapperEnvironment.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+public protocol MapperEnvironmentKey<Value>: Hashable {
+    associatedtype Value
+    static var defaultValue: Self.Value { get }
+}
+
+/// An environment dictionary that holds extra context in mappers
+public struct MapperEnvironment {
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/Mappers/MapperEnvironment.swift:9]: What about making this `Sendable` ensure the internal `environment` variable is accessed through a lock?

File: Sources/TuistCore/Graph/Mappers/MapperEnvironmentKeys.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+private struct InitialGraphKey: MapperEnvironmentKey {
+    static var defaultValue: Graph?
+}
+
+extension MapperEnvironment {
+    /// Hashes of targets that are missing in the remote storage
+    public var testsCacheUntestedHashes: [Target: String] {
+        get { self[TestsCacheHashesKey.self] }
+        set { self[TestsCacheHashesKey.self] = newValue }
+    }
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/Mappers/MapperEnvironmentKeys.swift:19]: This reads much better.

File: Sources/TuistCore/GraphTraverser/GraphTraversing.swift [hunk 3]
```diff
@@ -74,6 +76,8 @@ public protocol GraphTraversing {
     /// - Returns: All direct and transitive target dependencies
     func allTargetDependencies(path: AbsolutePath, name: String) -> Set<GraphTarget>
 
+    func allTargetDependencies(_ rootTargets: [GraphTarget]) -> Set<GraphTarget>
```
Comment (pepicrft) [inline Sources/TuistCore/GraphTraverser/GraphTraversing.swift:79]: Is this traversing from those targets? If so:
```suggestion
    func allTargetDependencies(traversingFromTargets: [GraphTarget]) -> Set<GraphTarget>
```

File: Sources/TuistCore/Graph/Mappers/GraphMapper.swift [hunk 4]
```diff
@@ -6,21 +6,21 @@ import XcodeGraph
 public protocol GraphMapping {
     /// Given a value graph, it maps it into another value graph.
     /// - Parameter graph: Graph to be mapped.
-    func map(graph: Graph) async throws -> (Graph, [SideEffectDescriptor])
+    func map(graph: Graph, environment: MapperEnvironment) async throws -> (Graph, [SideEffectDescriptor], MapperEnvironment)
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/Mappers/GraphMapper.swift:9]: In wonder if long-term we should evolve the API to be:
```suggestion
    func map(environment: MapperEnvironment) async throws -> (MapperEnvironment)
```

And merge graph and side effects into the environment.

**Review by pepicrft — APPROVED** [review]

File: Sources/TuistCore/Graph/Mappers/GraphMapper.swift (hunk 4 above)
Comment (fortmarek) [reply Sources/TuistCore/Graph/Mappers/GraphMapper.swift:9]: I think it's better to stay explicit about the inputs and outputs for now

File: Sources/TuistCore/Graph/Mappers/MapperEnvironment.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistCore/Graph/Mappers/MapperEnvironment.swift:9]: I don't follow why we need to make this `Sendable`. It's a struct passed around, not a single object.

**Pull request merged**